core/shared/src/main/scala/zio/ConfigProvider.scala

```diff
@@ -0,0 +1,351 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio
+
+import java.time.format.DateTimeFormatter
+
+/**
+ * A ConfigProvider is a service that provides configuration given a description
+ * of the structure of that configuration.
+ */
+trait ConfigProvider { self =>
+
+  /**
+   * Loads the specified configuration, or fails with a config error.
+   */
+  def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A]
+
+  /**
+   * Returns a new config provider that will automatically nest all
+   * configuration under the specified property name. This can be utilized to
+   * aggregate separate configuration sources that are all required to load a
+   * single configuration value.
+   */
+  final def nested(name: String): ConfigProvider =
+    new ConfigProvider {
+      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] = self.load(config.nested(name))
+    }
+
+  /**
+   * Returns a new config provider that preferentially loads configuration data
+   * from this one, but which will fall back to the specified alterate provider
+   * if there are any issues loading the configuration from this provider.
+   */
+  final def orElse(that: ConfigProvider): ConfigProvider =
+    new ConfigProvider {
+      def load[A](config: Config[A])(implicit trace: Trace): IO[Config.Error, A] =
+        self.load(config).orElse(that.load(config))
+    }
+}
+object ConfigProvider {
+
+  /**
+   * A simplified config provider that knows only how to deal with flat
+   * (key/value) properties. Because these providers are common, there is
+   * special support for implementing them.
+   */
+  trait Flat {
+    def load[A](path: Chunk[String], config: Config.Primitive[A])(implicit trace: Trace): IO[Config.Error, Chunk[A]]
+
+    def enumerateChildren(path: Chunk[String])(implicit trace: Trace): IO[Config.Error, Chunk[String]]
+  }
+  object Flat {
+    object util {
+      def parseAtom[A](
+        text: String,
+        path: Chunk[String],
+        name: String,
+        atom: Config.Primitive[A],
+        delim: String
+      ): IO[Config.Error, Chunk[A]] = {
+        val name    = path.lastOption.getOrElse("<unnamed>")
+        val unsplit = atom == Config.Secret
+
+        if (unsplit) ZIO.fromEither(atom.parse(text)).map(Chunk(_))
+        else
+          ZIO
+            .foreach(Chunk.fromArray(text.split("\\s*" + delim + "\\s*")))(s => ZIO.fromEither(atom.parse(s.trim)))
+            .mapError(_.prefixed(path))
+      }
```

Comment:
Would it be better to replace all `atom` to `primitive` like below?

```suggestion
      def parsePrimitive[A](
        text: String,
        path: Chunk[String],
        name: String,
        primitive: Config.Primitive[A],
        delim: String
      ): IO[Config.Error, Chunk[A]] = {
        val name    = path.lastOption.getOrElse("<unnamed>")
        val unsplit = primitive == Config.Secret

        if (unsplit) ZIO.fromEither(primitive.parse(text)).map(Chunk(_))
        else
          ZIO
            .foreach(Chunk.fromArray(text.split("\\s*" + delim + "\\s*")))(s => ZIO.fromEither(primitive.parse(s.trim)))
            .mapError(_.prefixed(path))
      }
```